File: dependencies/ann-hybrid/src/hnsw.rs
/// Hierarchical Navigable Small World (HNSW) graph for approximate nearest neighbor search.
///
/// Provides O(log n) semantic search over dense embedding vectors.

use rand::Rng;
use std::collections::BinaryHeap;
use std::cmp::Reverse;

/// A node in the HNSW graph.
#[derive(Debug, Clone)]
pub struct HnswNode {
    pub id: String,
    pub vector: Vec<f32>,
    /// Connections per layer: layer index → neighbor node indices
    pub connections: Vec<Vec<usize>>,
    pub level: usize,
}

/// Configuration for the HNSW index.
#[derive(Debug, Clone)]
pub struct HnswConfig {
    /// Maximum number of connections per node per layer.
    pub m: usize,
    /// Size of the dynamic candidate list during construction.
    pub ef_construction: usize,
    /// Size of the dynamic candidate list during search.
    pub ef_search: usize,
    /// Normalization factor for level generation.
    pub ml: f64,
}

impl Default for HnswConfig {
    fn default() -> Self {
        Self {
            m: 16,
            ef_construction: 200,
            ef_search: 50,
            ml: 1.0 / (16.0_f64).ln(),
        }
    }
}

/// The HNSW index.
pub struct HnswIndex {
    config: HnswConfig,
    nodes: Vec<HnswNode>,
    entry_point: Option<usize>,
    max_level: usize,
}

impl HnswIndex {
    pub fn new(config: HnswConfig) -> Self {
        Self {
            config,
            nodes: Vec::new(),
            entry_point: None,
            max_level: 0,
        }
    }

    /// Insert a vector with an ID into the index.
    pub fn insert(&mut self, id: &str, vector: &[f32]) {
        let level = self.random_level();
        let node_idx = self.nodes.len();

        let mut node = HnswNode {
            id: id.to_string(),
            vector: vector.to_vec(),
            connections: vec![Vec::new(); level + 1],
            level,
        };

        if self.nodes.is_empty() {
            self.nodes.push(node);
            self.entry_point = Some(0);
            self.max_level = level;
            return;
        }

        // Find entry point and greedily traverse from top layer
        let ep = self.entry_point.unwrap();
        let mut current = ep;

        // Traverse layers above node's level (greedy)
        for l in (level + 1..=self.max_level).rev() {
            current = self.greedy_closest(current, vector, l);
        }

        // For each layer from node's level down to 0, find and connect neighbors
        for l in (0..=level.min(self.max_level)).rev() {
            let neighbors = self.search_layer(current, vector, self.config.ef_construction, l);
            let selected: Vec<usize> = neighbors
                .into_iter()
                .take(self.config.m)
                .map(|(idx, _)| idx)
                .collect();

            node.connections[l] = selected.clone();

            // Add bidirectional connections
            for &neighbor_idx in &selected {
                if l < self.nodes[neighbor_idx].connections.len() {
                    self.nodes[neighbor_idx].connections[l].push(node_idx);
                    // Trim if over capacity
                    if self.nodes[neighbor_idx].connections[l].len() > self.config.m * 2 {
                        let nv = self.nodes[neighbor_idx].vector.clone();
                        self.nodes[neighbor_idx].connections[l]
                            .sort_by(|&a, &b| {
                                let da = cosine_distance(&nv, &self.nodes[a].vector);
                                let db = cosine_distance(&nv, &self.nodes[b].vector);
                                da.partial_cmp(&db).unwrap_or(std::cmp::Ordering::Equal)
                            });
                        self.nodes[neighbor_idx].connections[l].truncate(self.config.m);
                    }
                }
            }

            if !selected.is_empty() {
                current = selected[0];
            }
        }

        if level > self.max_level {
            self.max_level = level;
            self.entry_point = Some(node_idx);
        }

        self.nodes.push(node);
    }

    /// Search for the top-k nearest neighbors.
    pub fn search(&self, query: &[f32], k: usize) -> Vec<(String, f32)> {
        if self.nodes.is_empty() {
            return Vec::new();
        }

        let ep = self.entry_point.unwrap();
        let mut current = ep;

        // Traverse from top to layer 1 (greedy)
        for l in (1..=self.max_level).rev() {
            current = self.greedy_closest(current, query, l);
        }

        // Search layer 0 with ef_search candidates
        let candidates = self.search_layer(current, query, self.config.ef_search, 0);

        candidates
            .into_iter()
            .take(k)
            .map(|(idx, dist)| (self.nodes[idx].id.clone(), dist))
            .collect()
    }

    /// Number of items in the index.
    pub fn len(&self) -> usize {
        self.nodes.len()
    }

    pub fn is_empty(&self) -> bool {
        self.nodes.is_empty()
    }

    fn random_level(&self) -> usize {
        let mut rng = rand::thread_rng();
        let mut level = 0;
        while rng.gen::<f64>() < (1.0 / self.config.m as f64) && level < 16 {
            level += 1;
        }
        level
    }

    fn greedy_closest(&self, start: usize, query: &[f32], layer: usize) -> usize {
        let mut current = start;
        let mut current_dist = cosine_distance(query, &self.nodes[current].vector);

        loop {
            let mut changed = false;
            if layer < self.nodes[current].connections.len() {
                for &neighbor in &self.nodes[current].connections[layer] {
                    let dist = cosine_distance(query, &self.nodes[neighbor].vector);
                    if dist < current_dist {
                        current = neighbor;
                        current_dist = dist;
                        changed = true;
                    }
                }
            }
            if !changed {
                break;
            }
        }
        current
    }

    fn search_layer(
        &self,
        entry: usize,
        query: &[f32],
        ef: usize,
        layer: usize,
    ) -> Vec<(usize, f32)> {
        let mut visited = fnv::FnvHashSet::default();
        visited.insert(entry);

        let entry_dist = cosine_distance(query, &self.nodes[entry].vector);
        let mut candidates: BinaryHeap<Reverse<(OrderedFloat, usize)>> = BinaryHeap::new();
        let mut results: BinaryHeap<(OrderedFloat, usize)> = BinaryHeap::new();

        candidates.push(Reverse((OrderedFloat(entry_dist), entry)));
        results.push((OrderedFloat(entry_dist), entry));

        while let Some(Reverse((OrderedFloat(c_dist), c_idx))) = candidates.pop() {
            let worst_dist = results.peek().map(|(OrderedFloat(d), _)| *d).unwrap_or(f32::MAX);
            if c_dist > worst_dist && results.len() >= ef {
                break;
            }

            if layer < self.nodes[c_idx].connections.len() {
                for &neighbor in &self.nodes[c_idx].connections[layer] {
                    if visited.insert(neighbor) {
                        let dist = cosine_distance(query, &self.nodes[neighbor].vector);
                        let worst_dist = results.peek().map(|(OrderedFloat(d), _)| *d).unwrap_or(f32::MAX);

                        if dist < worst_dist || results.len() < ef {
                            candidates.push(Reverse((OrderedFloat(dist), neighbor)));
                            results.push((OrderedFloat(dist), neighbor));
                            if results.len() > ef {
                                results.pop();
                            }
                        }
                    }
                }
            }
        }

        let mut output: Vec<(usize, f32)> = results
            .into_iter()
            .map(|(OrderedFloat(d), idx)| (idx, d))
            .collect();
        output.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal));
        output
    }
}

/// Cosine distance = 1 - cosine_similarity.
pub fn cosine_distance(a: &[f32], b: &[f32]) -> f32 {
    let mut dot = 0.0f32;
    let mut norm_a = 0.0f32;
    let mut norm_b = 0.0f32;

    for i in 0..a.len().min(b.len()) {
        dot += a[i] * b[i];
        norm_a += a[i] * a[i];
        norm_b += b[i] * b[i];
    }

    let denom = norm_a.sqrt() * norm_b.sqrt();
    if denom < 1e-10 {
        return 1.0;
    }
    1.0 - (dot / denom)
}

/// Wrapper for f32 that implements Ord for use in BinaryHeap.
#[derive(Debug, Clone, Copy, PartialEq)]
struct OrderedFloat(f32);

impl Eq for OrderedFloat {}

impl PartialOrd for OrderedFloat {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.0.partial_cmp(&other.0)
    }
}

impl Ord for OrderedFloat {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.partial_cmp(other).unwrap_or(std::cmp::Ordering::Equal)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_vector(seed: f32, dim: usize) -> Vec<f32> {
        (0..dim).map(|i| ((seed + i as f32) * 0.1).sin()).collect()
    }

    #[test]
    fn test_hnsw_insert_and_search() {
        let mut index = HnswIndex::new(HnswConfig::default());
        for i in 0..50 {
            let v = make_vector(i as f32, 32);
            index.insert(&format!("item_{i}"), &v);
        }
        assert_eq!(index.len(), 50);

        let query = make_vector(5.0, 32);
        let results = index.search(&query, 5);
        assert!(!results.is_empty());
        assert!(results.len() <= 5);
        // item_5 should be the closest to itself
        assert_eq!(results[0].0, "item_5");
    }

    #[test]
    fn test_hnsw_empty() {
        let index = HnswIndex::new(HnswConfig::default());
        assert!(index.is_empty());
        let results = index.search(&[0.1, 0.2, 0.3], 5);
        assert!(results.is_empty());
    }

    #[test]
    fn test_cosine_distance() {
        let a = vec![1.0, 0.0, 0.0];
        let b = vec![1.0, 0.0, 0.0];
        assert!((cosine_distance(&a, &b) - 0.0).abs() < 1e-6);

        let c = vec![0.0, 1.0, 0.0];
        assert!((cosine_distance(&a, &c) - 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_hnsw_single_item() {
        let mut index = HnswIndex::new(HnswConfig::default());
        index.insert("only", &[1.0, 2.0, 3.0]);
        let results = index.search(&[1.0, 2.0, 3.0], 5);
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].0, "only");
    }
}

File: pkg/hnsw/hnsw.go
// Package hnsw implements Hierarchical Navigable Small World graphs for
// approximate nearest neighbor search on dense vectors.
package hnsw

import (
	"encoding/gob"
	"math"
	"math/rand"
	"os"
	"sort"
	"sync"
)

type DistanceFunc func(a, b []float32) float32

func CosineDistance(a, b []float32) float32 {
	var dot, normA, normB float32
	for i := range a {
		dot += a[i] * b[i]
		normA += a[i] * a[i]
		normB += b[i] * b[i]
	}
	if normA == 0 || normB == 0 {
		return 1
	}
	return 1 - dot/float32(math.Sqrt(float64(normA*normB)))
}

func L2Distance(a, b []float32) float32 {
	var sum float32
	for i := range a {
		d := a[i] - b[i]
		sum += d * d
	}
	return float32(math.Sqrt(float64(sum)))
}

type node struct {
	ID      string
	Vector  []float32
	Friends [][]string // friends[level] = list of neighbor IDs
}

type searchCandidate struct {
	ID   string
	Dist float32
}

type Index struct {
	mu             sync.RWMutex
	nodes          map[string]*node
	entryPoint     string
	maxLevel       int
	m              int     // max connections per layer
	mMax0          int     // max connections at layer 0
	efConstruction int
	ml             float64 // level multiplier
	dim            int
	dist           DistanceFunc
}

type Config struct {
	M              int
	EfConstruction int
	Dim            int
	Distance       string // "cosine" or "l2"
}

func New(cfg Config) *Index {
	if cfg.M == 0 {
		cfg.M = 16
	}
	if cfg.EfConstruction == 0 {
		cfg.EfConstruction = 200
	}
	dist := CosineDistance
	if cfg.Distance == "l2" {
		dist = L2Distance
	}
	return &Index{
		nodes:          make(map[string]*node),
		m:              cfg.M,
		mMax0:          cfg.M * 2,
		efConstruction: cfg.EfConstruction,
		ml:             1 / math.Log(float64(cfg.M)),
		dim:            cfg.Dim,
		dist:           dist,
		maxLevel:       -1,
	}
}

func (idx *Index) randomLevel() int {
	return int(math.Floor(-math.Log(rand.Float64()) * idx.ml))
}

func (idx *Index) Add(id string, vector []float32) {
	idx.mu.Lock()
	defer idx.mu.Unlock()

	level := idx.randomLevel()
	n := &node{
		ID:      id,
		Vector:  vector,
		Friends: make([][]string, level+1),
	}
	for i := range n.Friends {
		n.Friends[i] = []string{}
	}
	idx.nodes[id] = n

	if idx.maxLevel == -1 {
		idx.entryPoint = id
		idx.maxLevel = level
		return
	}

	ep := idx.entryPoint
	for l := idx.maxLevel; l > level; l-- {
		ep = idx.greedyClosest(vector, ep, l)
	}

	for l := min(level, idx.maxLevel); l >= 0; l-- {
		neighbors := idx.searchLayer(vector, ep, idx.efConstruction, l)

		maxConn := idx.m
		if l == 0 {
			maxConn = idx.mMax0
		}
		selected := idx.selectNeighbors(neighbors, maxConn)

		n.Friends[l] = make([]string, len(selected))
		for i, s := range selected {
			n.Friends[l][i] = s.ID
		}

		for _, s := range selected {
			neighbor := idx.nodes[s.ID]
			if l < len(neighbor.Friends) {
				neighbor.Friends[l] = append(neighbor.Friends[l], id)
				if len(neighbor.Friends[l]) > maxConn {
					cands := make([]searchCandidate, len(neighbor.Friends[l]))
					for i, fid := range neighbor.Friends[l] {
						cands[i] = searchCandidate{ID: fid, Dist: idx.dist(neighbor.Vector, idx.nodes[fid].Vector)}
					}
					trimmed := idx.selectNeighbors(cands, maxConn)
					neighbor.Friends[l] = make([]string, len(trimmed))
					for i, t := range trimmed {
						neighbor.Friends[l][i] = t.ID
					}
				}
			}
		}

		if len(neighbors) > 0 {
			ep = neighbors[0].ID
		}
	}

	if level > idx.maxLevel {
		idx.maxLevel = level
		idx.entryPoint = id
	}
}

func (idx *Index) Search(query []float32, k int, efSearch int) []SearchResult {
	idx.mu.RLock()
	defer idx.mu.RUnlock()

	if len(idx.nodes) == 0 {
		return nil
	}
	if efSearch == 0 {
		efSearch = max(k, 100)
	}

	ep := idx.entryPoint
	for l := idx.maxLevel; l > 0; l-- {
		ep = idx.greedyClosest(query, ep, l)
	}

	candidates := idx.searchLayer(query, ep, efSearch, 0)
	if len(candidates) > k {
		candidates = candidates[:k]
	}

	results := make([]SearchResult, len(candidates))
	for i, c := range candidates {
		results[i] = SearchResult{ID: c.ID, Distance: c.Dist}
	}
	return results
}

type SearchResult struct {
	ID       string
	Distance float32
}

func (idx *Index) Delete(id string) {
	idx.mu.Lock()
	defer idx.mu.Unlock()

	n, ok := idx.nodes[id]
	if !ok {
		return
	}
	for l, friends := range n.Friends {
		for _, fid := range friends {
			f := idx.nodes[fid]
			if l < len(f.Friends) {
				newFriends := make([]string, 0, len(f.Friends[l]))
				for _, ff := range f.Friends[l] {
					if ff != id {
						newFriends = append(newFriends, ff)
					}
				}
				f.Friends[l] = newFriends
			}
		}
	}
	delete(idx.nodes, id)

	if id == idx.entryPoint && len(idx.nodes) > 0 {
		for nid := range idx.nodes {
			idx.entryPoint = nid
			break
		}
	}
}

func (idx *Index) Len() int {
	idx.mu.RLock()
	defer idx.mu.RUnlock()
	return len(idx.nodes)
}

func (idx *Index) greedyClosest(query []float32, ep string, level int) string {
	best := ep
	bestDist := idx.dist(query, idx.nodes[ep].Vector)
	changed := true
	for changed {
		changed = false
		n := idx.nodes[best]
		if level < len(n.Friends) {
			for _, fid := range n.Friends[level] {
				d := idx.dist(query, idx.nodes[fid].Vector)
				if d < bestDist {
					bestDist = d
					best = fid
					changed = true
				}
			}
		}
	}
	return best
}

func (idx *Index) searchLayer(query []float32, ep string, ef int, level int) []searchCandidate {
	visited := map[string]bool{ep: true}
	candidates := []searchCandidate{{ID: ep, Dist: idx.dist(query, idx.nodes[ep].Vector)}}
	results := []searchCandidate{{ID: ep, Dist: idx.dist(query, idx.nodes[ep].Vector)}}

	for len(candidates) > 0 {
		sort.Slice(candidates, func(i, j int) bool { return candidates[i].Dist < candidates[j].Dist })
		closest := candidates[0]
		candidates = candidates[1:]

		worst := results[len(results)-1]
		if closest.Dist > worst.Dist && len(results) >= ef {
			break
		}

		n := idx.nodes[closest.ID]
		if level < len(n.Friends) {
			for _, fid := range n.Friends[level] {
				if visited[fid] {
					continue
				}
				visited[fid] = true
				d := idx.dist(query, idx.nodes[fid].Vector)

				if len(results) < ef || d < results[len(results)-1].Dist {
					candidates = append(candidates, searchCandidate{ID: fid, Dist: d})
					results = append(results, searchCandidate{ID: fid, Dist: d})
					sort.Slice(results, func(i, j int) bool { return results[i].Dist < results[j].Dist })
					if len(results) > ef {
						results = results[:ef]
					}
				}
			}
		}
	}
	return results
}

func (idx *Index) selectNeighbors(candidates []searchCandidate, m int) []searchCandidate {
	sort.Slice(candidates, func(i, j int) bool { return candidates[i].Dist < candidates[j].Dist })
	if len(candidates) > m {
		return candidates[:m]
	}
	return candidates
}

func (idx *Index) Save(path string) error {
	idx.mu.RLock()
	defer idx.mu.RUnlock()

	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	type exportData struct {
		Nodes      map[string]*node
		EntryPoint string
		MaxLevel   int
		M          int
		Dim        int
	}

	return gob.NewEncoder(f).Encode(exportData{
		Nodes:      idx.nodes,
		EntryPoint: idx.entryPoint,
		MaxLevel:   idx.maxLevel,
		M:          idx.m,
		Dim:        idx.dim,
	})
}

func (idx *Index) Load(path string) error {
	idx.mu.Lock()
	defer idx.mu.Unlock()

	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	type exportData struct {
		Nodes      map[string]*node
		EntryPoint string
		MaxLevel   int
		M          int
		Dim        int
	}

	var data exportData
	if err := gob.NewDecoder(f).Decode(&data); err != nil {
		return err
	}

	idx.nodes = data.Nodes
	idx.entryPoint = data.EntryPoint
	idx.maxLevel = data.MaxLevel
	idx.m = data.M
	idx.mMax0 = data.M * 2
	idx.dim = data.Dim
	return nil
}
